src/Storages/StorageKeeperMap.cpp

```diff
@@ -0,0 +1,694 @@
+    dropped_lock_path = fs::path(dropped_path) / "lock";
+
+    if (attach)
+    {
+        checkTable<false>();
+        return;
+    }
+
+    auto client = getClient();
+
+    if (root_path != "/" && !client->exists(root_path))
+    {
+        if (!create_missing_root_path)
+        {
+            throw Exception(
+                ErrorCodes::BAD_ARGUMENTS,
+                "Path '{}' doesn't exist. Please create it or set 'create_missing_root_path' to true'",
+                root_path_);
+        }
+        else
+        {
+            LOG_TRACE(log, "Creating root path {}", root_path);
+            client->createAncestors(root_path);
+            client->createIfNotExists(root_path, "");
+        }
+    }
+
+    for (size_t i = 0; i < 1000; ++i)
+    {
+        if (client->exists(dropped_path))
+        {
+            LOG_INFO(log, "Removing leftover nodes");
+            auto code = client->tryCreate(dropped_lock_path, "", zkutil::CreateMode::Ephemeral);
+
+            if (code == Coordination::Error::ZNONODE ||  code == Coordination::Error::ZNODEEXISTS)
+            {
+                LOG_INFO(log, "Someone else removed leftover nodes");
+            }
+            else if (code != Coordination::Error::ZOK)
+            {
+                throw Coordination::Exception(code, dropped_lock_path);
+            }
+            else
+            {
+                auto metadata_drop_lock = zkutil::EphemeralNodeHolder::existing(dropped_lock_path, *client);
+                if (!dropTable(client, metadata_drop_lock))
+                    continue;
+            }
+        }
+
+        client->createIfNotExists(metadata_path, "");
+        client->createIfNotExists(tables_path, "");
+
+        auto code = client->tryCreate(table_path, "", zkutil::CreateMode::Persistent);
+
+        if (code == Coordination::Error::ZOK)
+        {
+            table_is_valid = true;
+            return;
+        }
+
+        if (code == Coordination::Error::ZNONODE)
+            LOG_INFO(log, "Metadata nodes were deleted in background, will retry");
+        else
+            throw Coordination::Exception(code, table_path);
+    }
+
+    throw Exception(ErrorCodes::BAD_ARGUMENTS, "Cannot create metadata for table, because it is removed concurrently or because of wrong root_path ({})", root_path);
+}
+
+
+Pipe StorageKeeperMap::read(
+    const Names & column_names,
+    const StorageSnapshotPtr & storage_snapshot,
+    SelectQueryInfo & query_info,
+    ContextPtr context_,
+    QueryProcessingStage::Enum /*processed_stage*/,
+    size_t max_block_size,
+    unsigned num_streams)
+{
+    checkTable<true>();
+    storage_snapshot->check(column_names);
+
+    FieldVectorPtr filtered_keys;
+    bool all_scan;
+
+    Block sample_block = storage_snapshot->metadata->getSampleBlock();
+    auto primary_key_type = sample_block.getByName(primary_key).type;
+    std::tie(filtered_keys, all_scan) = getFilterKeys(primary_key, primary_key_type, query_info, context_);
```

Comment:
I've seen something similar in `StorageSystemZooKeeper`, maybe we can merge `extractPathImpl` and `getFilterKeys` into one generic function